pkg/config/expand.go

```diff
@@ -413,24 +413,52 @@ type varContext struct {
 	blueprint  Blueprint
 }
 
-func identifySimpleVariable(str string) (string, string, error) {
-	varComponents := strings.Split(str, ".")
+type varReference struct {
+	Group  string
+	Source string
+	Name   string
+}
 
-	// initialize these variables to empty string
-	var varSource, varValue string
-	// support only intra-group references of length 2
-	if len(varComponents) == 2 {
-		varSource = varComponents[0]
-		varValue = varComponents[1]
-	}
+/*
+This function performs only the most rudimentary conversion of an input
+string into a proper "variable reference". A variable reference has
+2 or 3 fields as described below:
+
+  - Source: a module ID or "vars" if referring to a Deployment Variable
+  - Group (optional): if the Source is a module ID, this must be the deployment
+    group in which the module is found. Optional when referring to a module
+    the same deployment group.
```

Comment:
Obvi some of these changes require some slightly more sophisticated processing of the string logic, but I think we should pick the data format based on what makes the most sense logically. I am not set in stone. You have put the most thought into how these will be used. Mostly just want to make sure the options were considered.

Reply:
Another thought: consider `vars` as an `id` from a special group called `top-level-vars`. I kind of like the idea of all three value being required.

Reply:
I prefer to keep group have a very tight coupling to deployment groups. Since `vars` is not a a deployment group, I prefer not to have it have a double meaning. But I could imagine changes like the following:

`SourceID`: must be "vars" or a "module ID"
`GroupID`: allowed to be empty string except when `SourceID` is "vars" in which case it will be set to "global" (or similar choice)
`Name`: the name of the deployment variable or module output

A problem with this is that we'd also have to enforce that a deployment group cannot be named "global". Likewise, we might want to reject modules named "vars". What do you think of that as a separate task? With that done, we could make this simple change.

I suppose the question here is whether you think the ID of the current group ought to be supplied to this function so that `GroupID` is always populated with a value. For the next step of this code, I don't think this change buys us all that much., Since module IDs are unique it is already easy to detect an intergroup reference. The status quo is to fail when the user does this.

In our design, one of the things we want to enforce is "if the module is in another group, did the user explicitly identify the group?" Which is a step best left for a function more purely dedicated to testing the validity (rather than syntax) of a reference.

If we find that the use of this struct starts expanding and it would be useful for GroupID to always be populated, I think it would be not too hard to add a "source group ID" argument to this function and ensure that `GroupID` is always populated.

But I'd like to stick to the field names and behavior suggested above.

Reply:
+1 on not making vars a group
+1 on making sure someone does not name a group `vars`. Not sure if we should also ban `global`, but definitely if we are going to make group mandator and have a static group for vars.
+1 to having group always populated (except maybe for vars, not sure what is best there, blank or global). 

Seems like outcome is: 
- Separate task to ban some group names (such as `vars`)
- Separate task to add `source group ID` as arg (or were you proposing to do that here?). I don't think it is needed until we start using this object elsewhere in the code.

Reply:
In an offline we discussion, we decided best approach anticipating some future needs is to populate `GroupID` on every call and to record whether it was populated "explicitly" (through the reference itself) or "implicitly" by the using the group ID of where the reference was located. This update is on the present head of the feature branch for one more review.